src/duration.cpp

```diff
@@ -0,0 +1,208 @@
+#include <chrono>
+#include <cstdint>
+#include <sstream>
+#include <iomanip>
+#include <regex>
+#include <stdexcept>
+#include <tuple>
+#include <limits>
+#include <Rcpp.h>
+#include "duration.hpp"
+#include "pseudovector.hpp"
+#include "utilities.hpp"
+
+Global::duration from_string(const std::string& str) {
+  Global::duration d = std::chrono::seconds(0);
+  const char* s = str.c_str();
+  const char* e = s + str.size();
+
+  auto sign = 1;
+  if (s < e && *s == '-') {
+    sign = -1;
+    ++s;
+  }
+
+  int n;
+  if (!Global::readNumber(s, e, n, false)) throw std::range_error("cannot parse nanoduration");
+
+  if (s < e && *s == ':') {
+    // we've got HHH:MM:SS format
+    d += n * std::chrono::hours(1);
+    ++s;
+    if (s + 5 > e || !isdigit(s[0]) || !isdigit(s[1]) || 
+        s[2] != ':' || !isdigit(s[3]) || !isdigit(s[4])) {
+      throw std::range_error("cannot parse nanoduration");
```

Comment:
I am wondering whether this file should 

- get a `try/catch` block it currently does not have
- use `Rcpp::stop()` instead of `throw` 

The preference for `Rcpp::stop()` would apply to the other files too, of course.

Reply:
Not sure. This is a low level function with no call to any of the R interface and when it is called it is protected by a try/catch block in `RcppExport SEXP duration_from_string(SEXP s)`. You could see it as a library function and like the STL functions, it  can throw. But one has to believe the exception mechanism will handle things correctly or else one would never use any C++: more specifically, after the catch in `duration_from_string` we have to be in a correct state or else C++ is broken (and we're not in a weird case like throwing from inside a constructor with partially allocated stuff).

On the other hand, I'm not sure of the subtleties before returning to R. So maybe `Rcpp::stop` should replace the `forward_exception_to_r`? I will look into the source code of `Rcpp::stop` to see what it's doing and get a better feel for some of the issues we could encounter. We could certainly replace all the `forward_expection_to_r` if `Rcpp::stop` is better.

Reply:
It's complicated.  You and I know the code will only be called from out try/catch.  But as it stands, it is a naked throw.  Which is a minor smell.  But heck, as we said, "we know" and we could just add a line of documentation.

As for the `Rcpp::stop()`  -- I tend to use it everywhere (when Rcpp is used) instead of a plain `throw` because it has some associated code with it that should generally be beneficial and not impose cost. Call it style preference.  _Effectively_ the difference may not matter.  Just how I tend to indent C/C++/R code by four spaces but everybody using RStudio comes at me with two spaces.  Oh well  :)